File: apps/android/app/src/main/java/ai/openclaw/android/node/SignificantLocationMonitor.kt
package ai.openclaw.android.node

import ai.openclaw.android.LocationMode
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject

/**
 * Monitors significant location changes (~500m displacement) and pushes
 * `location.update` events to the gateway so the severance hook can
 * determine whether the user is at their configured work location.
 */
object SignificantLocationMonitor {
  fun startIfNeeded(
    scope: CoroutineScope,
    location: LocationCaptureManager,
    locationMode: StateFlow<LocationMode>,
    hasFineLocationPermission: () -> Boolean,
    hasCoarseLocationPermission: () -> Boolean,
    sendNodeEvent: suspend (event: String, payloadJson: String) -> Unit,
  ) {
    if (locationMode.value == LocationMode.Off) return
    if (!hasFineLocationPermission() && !hasCoarseLocationPermission()) return
    location.startMonitoringSignificantChanges { lat, lon, accuracyMeters ->
      scope.launch {
        sendNodeEvent(
          "location.update",
          buildJsonObject {
            put("lat", JsonPrimitive(lat))
            put("lon", JsonPrimitive(lon))
            put("accuracyMeters", JsonPrimitive(accuracyMeters.toDouble()))
            put("source", JsonPrimitive("android-significant-location"))
          }.toString(),
        )
      }
    }
  }
}

File: apps/ios/Sources/Location/SignificantLocationMonitor.swift
import CoreLocation
import Foundation
import OpenClawKit

/// Monitors significant location changes and pushes `location.update`
/// events to the gateway so the severance hook can determine whether
/// the user is at their configured work location.
@MainActor
final class SignificantLocationMonitor {
    private let locationService: any LocationServicing
    private let locationMode: () -> OpenClawLocationMode
    private let sendEvent: @Sendable (String, String) async -> Void

    init(
        locationService: any LocationServicing,
        locationMode: @escaping () -> OpenClawLocationMode,
        sendEvent: @escaping @Sendable (String, String) async -> Void
    ) {
        self.locationService = locationService
        self.locationMode = locationMode
        self.sendEvent = sendEvent
    }

    func start() {
        let mode = self.locationMode()
        guard mode == .always else { return }
        let status = self.locationService.authorizationStatus()
        guard status == .authorizedAlways else { return }
        self.locationService.startMonitoringSignificantLocationChanges { [weak self] location in
            guard self != nil else { return }
            struct Payload: Codable {
                var lat: Double
                var lon: Double
                var accuracyMeters: Double
                var source: String?
            }
            let payload = Payload(
                lat: location.coordinate.latitude,
                lon: location.coordinate.longitude,
                accuracyMeters: location.horizontalAccuracy,
                source: "ios-significant-location")
            guard let data = try? JSONEncoder().encode(payload),
                  let json = String(data: data, encoding: .utf8)
            else { return }
            Task { @MainActor in
                await self?.sendEvent("location.update", json)
            }
        }
    }
}
